### PR TITLE
php83: init at 8.3.0alpha2

### PR DESCRIPTION
**Pull request by drupol**
This PR:

- Introduces PHP 8.3.0 at version alpha2
- Includes a patch from @nielsdos (_thanks mate!_) for fixing the issue with fileinfo extension (see https://github.com/php/php-src/issues/11408)
- Includes a patch for APCU extension from https://github.com/krakjoe/apcu/issues/489
- Includes a patch for session extension from https://github.com/php/php-src/pull/11530

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

File: pkgs/development/interpreters/php/generic.nix [hunk 1]
```diff
@@ -33,7 +33,7 @@ let
     , jq
 
     , version
-    , hash
+    , hash ? ""
```
Comment (jtojnar) [inline pkgs/development/interpreters/php/generic.nix:36]: Why not just pass empty hash up there? This will make it easier to accidentally omit.

Comment (drupol) [reply pkgs/development/interpreters/php/generic.nix:36]: This is a totally opinionated change.

File: pkgs/development/interpreters/php/8.3.nix [hunk 2]
```diff
@@ -0,0 +1,57 @@
+{ callPackage, fetchurl, ... }@_args:
+
+let
+  version = "8.3.0";
+  base = (callPackage ./generic.nix (_args // {
+    version = "8.3.0";
```
Comment (jtojnar) [inline pkgs/development/interpreters/php/8.3.nix:6]: The version should include the alpha status.

File: pkgs/development/interpreters/php/generic.nix (hunk 1 above)
Comment (drupol) [reply pkgs/development/interpreters/php/generic.nix:36]: This is an opinionated change, I just found it odd to have an empty hash in the file 8.3.nix.

Anyway, if the `hash` value is empty, will nix will complain when building, so, this is the reason.

File: pkgs/development/interpreters/php/8.3.nix [hunk 3]
```diff
@@ -0,0 +1,57 @@
+{ callPackage, fetchurl, ... }@_args:
+
+let
+  version = "8.3.0";
```
Comment (jtojnar) [inline pkgs/development/interpreters/php/8.3.nix:4]: Unused.

File: pkgs/development/interpreters/php/8.3.nix (hunk 2 above)
Comment (drupol) [reply pkgs/development/interpreters/php/8.3.nix]: Fixed.

File: pkgs/development/interpreters/php/8.3.nix (hunk 3 above)
Comment (drupol) [reply pkgs/development/interpreters/php/8.3.nix]: Fixed.

File: pkgs/development/interpreters/php/generic.nix (hunk 1 above)
Comment (jtojnar) [reply pkgs/development/interpreters/php/generic.nix:36]: You could just set it to `null`. And having the `hash` argument optional will fail earlier, saving developer’s time, and produce a more informative trace.

File: pkgs/development/interpreters/php/8.3.nix [hunk 4]
```diff
@@ -0,0 +1,56 @@
+{ callPackage, fetchurl, ... }@_args:
+
+let
+  base = (callPackage ./generic.nix (_args // {
+    version = "8.3.0-alpha2";
+  })).overrideAttrs (oldAttrs: {
+    src = (fetchurl {
```
Comment (jtojnar) [inline pkgs/development/interpreters/php/8.3.nix:7]: Parens not needed.

Comment (drupol) [reply pkgs/development/interpreters/php/8.3.nix]: Fixed.

File: pkgs/development/interpreters/php/generic.nix (hunk 1 above)
Comment (jtojnar) [reply pkgs/development/interpreters/php/generic.nix:36]: The default value is still here.

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

File: pkgs/top-level/php-packages.nix [hunk 5]
```diff
@@ -332,7 +332,12 @@ lib.makeScope pkgs.newScope (self: with self; {
         }
         { name = "exif"; doCheck = false; }
         { name = "ffi"; buildInputs = [ libffi ]; }
-        { name = "fileinfo"; buildInputs = [ pcre2 ]; }
+        {
+          name = "fileinfo";
+          buildInputs = [ pcre2 ];
+          patches = lib.optionals (lib.versionAtLeast php.version "8.3")
+            [ ../development/interpreters/php/fix-fileinfo-ext-php83.patch ];
```
Comment (jtojnar) [inline pkgs/top-level/php-packages.nix:339]: Would be nice to include a comment with explanation and source too:


```suggestion
          patches = lib.optionals (lib.versionAtLeast php.version "8.3") [
            # Fix the extension unable to be loaded due to missing `get_module` function.
            # `ZEND_GET_MODULE` macro that creates it is conditional on `COMPILE_DL_FILEINFO` being defined.
            # https://github.com/php/php-src/issues/11408#issuecomment-1602106200
            ../development/interpreters/php/fix-fileinfo-ext-php83.patch
          ];
```

Comment (drupol) [reply pkgs/top-level/php-packages.nix:339]: I'm pretty sure this will be fixed in the next alpha, but ok, I'll take care of this after lunch

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

File: pkgs/top-level/php-packages.nix [hunk 6]
```diff
@@ -556,7 +560,8 @@ lib.makeScope pkgs.newScope (self: with self; {
         { name = "tidy"; configureFlags = [ "--with-tidy=${html-tidy}" ]; doCheck = false; }
         {
           name = "tokenizer";
-          patches = [ ../development/interpreters/php/fix-tokenizer-php81.patch ];
+          patches = lib.optionals (lib.versionAtLeast php.version "8.1")
```
Comment (jtojnar) [inline pkgs/top-level/php-packages.nix:563]: This is redundant now that PHP 8.0 is not supported.

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

File: pkgs/top-level/php-packages.nix [hunk 7]
```diff
@@ -625,5 +629,7 @@ lib.makeScope pkgs.newScope (self: with self; {
       # Produce the final attribute set of all extensions defined.
     in
     builtins.listToAttrs namedExtensions
-  );
+  ) // lib.optionalAttrs (!(lib.versionAtLeast php.version "8.3")) {
```
Comment (jtojnar) [inline pkgs/top-level/php-packages.nix:632]: You can use `lib.versionOlder`

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

**Comment by etu** [comment]
So PHP builds fine, the FPM test works fine... However, Apache is quite
broken.

I just get this stack trace: http://ix.io/4z3v ending in a core dump. To me
this seems like an upstream issue, however I'm not sure about that. It's
just a hunch.

**Comment by jtojnar** [comment]
@etu see https://github.com/php/php-src/issues/11529

File: pkgs/development/interpreters/php/fix-fileinfo-ext-php83.patch [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+@@ -14,6 +14,9 @@
+   +----------------------------------------------------------------------+
+ */
+
++#ifdef HAVE_CONFIG_H
++#include "config.h"
++#endif
+ #include "php.h"
+
+ #include "libmagic/magic.h"
```
Comment (etu) [inline pkgs/development/interpreters/php/fix-fileinfo-ext-php83.patch:14]: I would like to have a newline added here to get rid of the missing newline symbol

Comment (drupol) [reply pkgs/development/interpreters/php/fix-fileinfo-ext-php83.patch]: Done

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

**Comment by drupol** [comment]
@GrahamcOfBorg build php81 php82 php83
@GrahamcOfBorg test php81 php82 php83

**Comment by drupol** [comment]
@GrahamcOfBorg test php81 php82 php83

**Review by etu — APPROVED** [review]
Well, now the tests pass and all that. I think it looks good. Also awesome that we managed to help to squash some bugs upstream while at it :smiley:

**Pull request merged**